### PR TITLE
Improve error message while setting error twice.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54464 Improve error message while setting error twice.**

In case where we accidentaly set an error twice on a Future, we get a
cryptic error like this:

```
Exception in thread pool task: !completed() INTERNAL ASSERT FAILED at "aten/src/ATen/core/ivalue_inl.h":534, please report a bug to PyTorch.
```

This PR, updates the error message to include some additional information about
what the previous error was.

Differential Revision: [D27249758](https://our.internmc.facebook.com/intern/diff/D27249758/)

### CODE REVIEW

File: test/cpp/jit/test_misc.cpp [hunk 1]
```diff
@@ -2203,6 +2204,17 @@ TEST(FuturesTest, CollectAny) {
   ASSERT_EQ(c4->value().toInt(), 7);
 }
 
+TEST(FuturesTest, SetError) {
```
Comment (swolchok) [inline test/cpp/jit/test_misc.cpp:2207]: shouldn't this go in aten/src/ATen/test/ivalue_test.cpp?

File: aten/src/ATen/core/ivalue_inl.h [hunk 2]
```diff
@@ -558,7 +558,11 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
   void setErrorInternal(
       std::exception_ptr eptr,
       std::unique_lock<std::mutex>& lock) {
-    AT_ASSERT(!completed());
+    TORCH_CHECK(
+        !eptr_,
+        "Error already set on the Future: ",
```
Comment (swolchok) [inline aten/src/ATen/core/ivalue_inl.h:563]: nit: s/the Future/this Future/ ?

**Review by swolchok — CHANGES_REQUESTED** [review]
setErrorInternal looks fine to me, though I think @ezyang has more specific thoughts on error message wording than I do. please move the test

**Comment by ezyang** [comment]
> though I think @ezyang has more specific thoughts on error message wording than I do

Wording seems fine, don't think this will actually show up for end users

File: aten/src/ATen/test/ivalue_test.cpp [hunk 3]
```diff
@@ -308,6 +309,18 @@ TEST(IValueTest, FutureExceptions) {
...
 }
 
+TEST(IValueTest, FutureSetError) {
+  auto f1 = c10::make_intrusive<ivalue::Future>(IntType::get());
+  f1->setError(std::make_exception_ptr(std::runtime_error("foo")));
+  try {
+    f1->setError(std::make_exception_ptr(std::runtime_error("bar")));
+    FAIL() << "Expected to throw";
+  } catch (std::exception& e) {
+    EXPECT_THAT(e.what(), ::testing::HasSubstr("Error already set"));
```
Comment (rohan-varma) [inline aten/src/ATen/test/ivalue_test.cpp:319]: nit: also test that it has the original error message? 
EXPECT_THAT(e.what(), ::testing::HasSubstr("foo"));

**Comment by pritamdamania87** [comment]
@swolchok Could you take another look at this PR? Thanks!

**Review by swolchok — APPROVED** [review]
Looks good, sorry for the delay!

**Pull request closed without merging**